digitalearthau/coherence.py

```diff
@@ -40,63 +47,121 @@ def main(expressions, archive_locationless, debug, check_locationless, check_anc
+    with Datacube(config=config_file) as dc:
         _LOG.info('query', query=expressions)
         count = 0
         archive_count = 0
+        locationless_count = 0
         for dataset in dc.index.datasets.search(**expressions):
             count += 1
             # Archive if it has no locations.
             # (the sync tool removes locations that don't exist anymore on disk,
             # but can't archive datasets as another path may be added later during the sync)
             if check_locationless or archive_locationless:
                 if dataset.uris is not None and len(dataset.uris) == 0:
-                    _LOG.info("dataset.locationless", dataset_id=str(dataset.id))
+                    locationless_count += 1
+
                     if archive_locationless:
                         dc.index.datasets.archive([dataset.id])
                         archive_count += 1
+                        _LOG.info("locationless_dataset_id.archived", dataset_id=str(dataset.id))
+                    else:
+                        _LOG.info("locationless_dataset_id", dataset_id=str(dataset.id))
 
             # If an ancestor is archived, it may have been replaced. This one may need
             # to be reprocessed too.
-            if check_ancestors:
-                _check_ancestors(check_siblings, dc, dataset)
+            if check_ancestors or archive_siblings or check_siblings:
+                archive_count += _check_ancestors(check_ancestors, check_siblings, archive_siblings, dc, dataset)
+
+        _LOG.info("coherence.finish",
+                  datasets_count=count,
+                  locationless_count=locationless_count,
+                  siblings_count=_siblings_count,
+                  archived_count=archive_count)
+
 
-        _LOG.info("coherence.finish", count=count, archive_count=archive_count)
+def _archive_dataset_ids(dc, ds_id, sibling_ids):
+    # sibling/s indexed_time dictionary
+    sb_itime_dict = dict()
 
+    # Count of dataset id's archived
+    archive_count = 0
 
-def _check_ancestors(check_siblings: bool, dc: Datacube, dataset: Dataset):
+    # dataset indexed time
+    ds_itime = dc.index.datasets.get(ds_id).indexed_time
+    for key, s_id in enumerate(sibling_ids):
+        sb_itime_dict[s_id] = dc.index.datasets.get(s_id).indexed_time
+
+    # new dataset id to retain
+    new_id = ds_id
+    for id, itime in sb_itime_dict.items():
+        # Check if dataset indexed time is less than or equal to sibling dataset indexed time
+        if ds_itime <= itime:
+            ds_id_archive = new_id
+            ds_archive_time = ds_itime
+
+            # Latest sibling dataset id to retain
+            new_id = id
+            ds_itime = itime
+        else:
+            ds_id_archive = id
+            ds_archive_time = itime
+
+        dc.index.datasets.archive([ds_id_archive])
+        _LOG.info("\tarchived_dataset_id", id=str(ds_id_archive), indexed_time=str(ds_archive_time))
```

Comment:
The `_LOG` in this module is a structured logger, which is intended to be able to output logs in an easily machine readable format. It would be better not to use `\t` or similar characters in the *key* of the structured log message.

Reply:
This code is replaced by new function _archive_duplicate_siblings.